types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+
+	wl_list_init(&connector->resources);
+	wl_list_insert(&device->connectors, &connector->link);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		drm_lease_connector_v1_send_to_client(
+				connector, wl_resource_get_client(resource), resource);
+	}
+
+	return true;
+}
+
+
+void wlr_drm_lease_manager_withdraw_output(
+	struct wlr_drm_lease_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+
+	wlr_log(WLR_DEBUG, "Withdrawing output %s", output->name);
+
+	struct wlr_drm_connector *drm_connector =
+			get_drm_connector_from_output(output);
+
+	struct wlr_drm_lease_device_v1 *device = NULL;
+	wl_list_for_each(device, &manager->devices, link) {
+		struct wlr_drm_backend *backend = get_drm_backend_from_backend(
+				device->backend);
+		if (backend == drm_connector->backend) {
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"given output");
+		return;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector = NULL, *tmp_conn;
+	wl_list_for_each(tmp_conn, &device->connectors, link) {
+		if (tmp_conn->output == output) {
+			connector = tmp_conn;
+			break;
+		}
+	}
+
+	if (!connector) {
+		wlr_log(WLR_DEBUG, "Output has not been offered");
+		return;
+	}
+
+	if (connector->active_lease) {
+		wlr_drm_lease_v1_revoke(connector->active_lease);
+	}
+
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &connector->resources) {
+		wp_drm_lease_connector_v1_send_withdrawn(resource);
+		wl_resource_set_user_data(resource, NULL);
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_list_init(wl_resource_get_link(resource));
+	}
+
+	wl_resource_for_each(resource, &device->requests) {
+		struct wlr_drm_lease_request_v1 *request =
+			wlr_drm_lease_request_v1_from_resource(resource);
+		request->invalid = true;
+	}
+
+	wl_list_remove(&connector->link);
+	wl_list_init(&connector->link);
+	free(connector);
+}
+
+static void wlr_drm_lease_device_v1_destroy(
+		struct wlr_drm_lease_device_v1 *device) {
+	struct wlr_drm_backend *backend =
+			get_drm_backend_from_backend(device->backend);
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_device_v1 for %s",
+			backend->name);
+
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->requests) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->leases) {
+		struct wlr_drm_lease_v1 *lease =
+			wlr_drm_lease_v1_from_resource(resource);
+		wlr_drm_lease_v1_revoke(lease);
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector, *tmp;
+	wl_list_for_each_safe(connector, tmp, &device->connectors, link) {
+		wl_list_remove(&connector->link);
+		wl_list_init(&connector->link);
+		free(connector);
+	}
+
+	free(device);
```

Comment:
We need to destroy the global here. But we need to be careful wrt. race conditions: use the `wlr_global_destroy_safe` function.